active/0000-building-target-assets.md

```diff
@@ -0,0 +1,187 @@
+```
+var app = new EmberApp(defaults, {
+  minifyJS: {
+    enabled: false
+  },
+  minifyCSS: {
+    enabled: false
+  },
+  additionalTargets: {
+    "fastboot": {
+      "dependsOn": ["browser"]
+    }
+  }
+});
+
+//...
+return app.toTree();
+```
+
+## Additional options to existing CLI commands
+The existing ember-cli commands only assume there is one target that it needs to build/generate files for. With the concept of targets, we now want to now allow the CLI commands to allow targets to be generated and created. The following will be the workflow for a ember user intending to create and use target.
+
+1. A user can generate a target using the following command: `ember generate target <target name>`
+This will end up creating a `fastboot` directory with app folder inside if it is done in an app or app and addon directories if done in an addon. Note: The target name should be the same as the one registered by the app/addon.
+
+2. A user can generate a route in a target environment: `ember generate route <route name> --target <target name>`
+This generates a route in the targeted environment.
+
+3. Building a target: `ember build --target <target name>`
+This will build the target artifacts. If the target depends on other targets, those will also be built as part of this command. The generated assets will all be under `dist/assets` with the following file naming:
+```
+dist/
+|-- assets/
+│   |-- vendor.js
+│   |-- app.js
+│   |-- vendor-<targetname>.js
+│   |-- app-<targetname>.js
+```
+
+Note:
+1. The command `ember serve` will remain as it is today since it serves the assets based on what is requested. If the targeted assets need to be served, the app/addon should update the index.html or their environment in which this is loaded to be served. The targeted app and addon directories should be watched so that a change in them triggers a rebuild.
+2. If a target is not defined and a user uses the above command, we will throw a self explanatory error message.
+3. `vendor-<targetname>.js` file will only contain the AMD modules for building addons and not the ember-cli/ember.js trees. These are already present in the vendor file and we expect users of this feature to be using the additional targets as a compliment to the existing ones.
+
+## Public hooks
+`ember-cli` as part of the build pipeline exposes public API for addons to change the different trees and processing per their needs. Since the targeted builds are doing an additional builds as a compliment to the existing browser build, we need to expose similar APIs for the target builds as well. Following are the public apis we need to expose:
+
+- `preProcessTree(type, tree)`: The existing type for browser build is represented by `javascript`, `styles` etc. We need to expose additional types when targetted builds are run. Therefore, for a target build the type will be invoked as `javascript-<targetname>`, `styles-<targetname>` etc.
```

Comment:
I am not sure I follow this. I am not changing the existing API but adding more types.